filebrowser/rockcraft.yaml

```diff
@@ -0,0 +1,158 @@
+    override-build: |
+      mkdir -p ${CRAFT_PART_INSTALL}/usr/share/rocks
+      (echo "# os-release" && cat /etc/os-release && echo "# dpkg-query" && dpkg-query -f '${db:Status-Abbrev},${binary:Package},${Version},${source:Package},${Source:Version}\n' -W) > ${CRAFT_PART_INSTALL}/usr/share/rocks/dpkg.query
+
+  healthcheck:
+    plugin: nil
+    source: https://github.com/filebrowser/filebrowser.git
+    source-tag: v2.27.0
+    override-build: |
+      cp healthcheck.sh $CRAFT_PART_INSTALL/healthcheck.sh
+    stage-packages:
+      - curl
+      - jq
+    permissions:
+      - path: healthcheck.sh
+        # 584792 is the _daemon_ user
+        owner: 584792
+        group: 584792
+        mode: "755"
+
+  add-config:
+    plugin: nil
+    source: https://github.com/filebrowser/filebrowser.git
+    source-tag: v2.27.0
+    build-packages:
+      - jq
+    override-build: |
+      jq '.database = "/tmp/database.db"' docker_config.json > $CRAFT_PART_INSTALL/.filebrowser.json
+    # TODO: Omitting stage here means .filebrowser.json does not get copied automatically.  Not sure why
+    #       Asked here: https://matrix.to/#/!NPPCseDHKRvSBMUEXN:ubuntu.com/$3pkoT_gtSQlIa_J2lv6iy3P0JdzMQ5Cbt38guXG9Sg4?via=ubuntu.com&via=matrix.org
+    stage:
+      - ".filebrowser.json"
+
+  frontend:
+    # Tried to use `plugin: npm` but it does:
+    #  * downloads node
+    #  * `npm pack`
+    # `npm pack` only tars the dist, it does not actually build anything.  We
+    # need to run `npm run build`.  Instead, we use `plugin: nil` and download
+    # node manually.
+    # plugin: npm
+    # npm-include-node: true
+    # npm-node-version: "18.19.0"
+    plugin: nil
+    source: https://github.com/filebrowser/filebrowser.git
+    source-tag: v2.27.0
+    # tried to use source-subdir, but override-build ignores it.  Maybe source-subdir is
+    # implemented as part of the default build step?
+    # source-subdir: frontend
+    build-environment:
+      # Set the node version for build-override here.
+      - NODE_VERSION: '18.18.2'
+    override-build: |
+      cd frontend
+
+      # Setup node/npm
+      node_uri="https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-x64.tar.gz"
+      echo fetching node for build: $node_uri
+      # Download and unpack right here
+      curl $node_uri | tar xzf - -C ./ --no-same-owner --strip-components=1
+      PATH=$(pwd)/bin:$PATH
+
+      # Build the frontend
+      npm ci
+      npm run build
+      # TODO: Is there a native way to do this?  Or do I lose this behaviour by 
+      # doing override-build?
+      mkdir -p $CRAFT_PART_INSTALL/frontend
+      cp -r $CRAFT_PART_BUILD/frontend/dist $CRAFT_PART_INSTALL/frontend
+    # We do not want anything from this part to get primed as it is only as a build artifact
+    # for the backend part.  By default, `prime` copies everything from stage to prime, so we
+    # override `prime` here to negative match all files and ensure it copies nothing.
```

Comment:
:exploding_head:

Reply:
don't get me started, @orfeas-k :laughing:

Reply:
I wonder if, for things like this that we'll likely repeat but don't want to forget it was intentional, we should have some github issues or something else we could link.  So instead of this explanation, it could just be:
```
# Override prime per http://github.com/canonical/something/issues/1234
prime: 
  - "-*"
```

it would also give us something easy to search for if we ever wanted to find and replace all our patches like this

Reply:
wdyt?  could also be linking to a style guide spec with a linkable bookmark on each item, although those links get LONG :D

Reply:
Good points, both of these. I think having a style guide is needed more and more as we create more ROCKs, but maybe until we actually create one, we can have issues (which could also be closed) that document stylistic decisions or behaviours we need to implement in order patch rockcraft default behaviour. On the long links point, you 're right but it's not a blocker I guess.

Reply:
k ill resolve this for now, but add a note to our rockcraft meeting